src/transformers/models/musicgen/modeling_musicgen.py

```diff
@@ -0,0 +1,1917 @@
+        self.num_codebooks = config.num_codebooks
+        self.embed_scale = math.sqrt(config.d_model) if config.scale_embedding else 1.0
+
+        embed_dim = config.vocab_size + 1
+        self.embed_tokens = nn.ModuleList(
+            [nn.Embedding(embed_dim, config.d_model) for _ in range(config.num_codebooks)]
+        )
+
+        self.embed_positions = MusicgenSinusoidalPositionalEmbedding(
+            config.max_position_embeddings,
+            config.d_model,
+        )
+
+        self.layers = nn.ModuleList([MusicgenDecoderLayer(config) for _ in range(config.num_layers)])
+        self.layer_norm = nn.LayerNorm(config.d_model)
+
+        self.gradient_checkpointing = False
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    # Copied from transformers.models.bart.modeling_bart.BartDecoder._prepare_decoder_attention_mask
+    def _prepare_decoder_attention_mask(self, attention_mask, input_shape, inputs_embeds, past_key_values_length):
+        # create causal mask
+        # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+        combined_attention_mask = None
+        if input_shape[-1] > 1:
+            combined_attention_mask = _make_causal_mask(
+                input_shape,
+                inputs_embeds.dtype,
+                device=inputs_embeds.device,
+                past_key_values_length=past_key_values_length,
+            )
+
+        if attention_mask is not None:
+            # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+            expanded_attn_mask = _expand_mask(attention_mask, inputs_embeds.dtype, tgt_len=input_shape[-1]).to(
+                inputs_embeds.device
+            )
+            combined_attention_mask = (
+                expanded_attn_mask if combined_attention_mask is None else expanded_attn_mask + combined_attention_mask
+            )
+
+        return combined_attention_mask
+
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        encoder_hidden_states: Optional[torch.FloatTensor] = None,
+        encoder_attention_mask: Optional[torch.LongTensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        cross_attn_head_mask: Optional[torch.Tensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, BaseModelOutputWithPastAndCrossAttentions]:
+        r"""
+        Args:
+            input_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
```

Comment:
The shape is incorrect no?

Reply:
Shouldn't it be `(batch_size, num_codebook_vectors, sequence_length)`?

Reply:
I think it'd actually be better to change the logic in the code though so that the shape is `(batch_size * num_codebook_vectors, sequence_length)` instead as stated below.

Reply:
This could get messy since we sum the `input_embeds` over codebooks (so we'd need to reshape it again from `batch_size * num_codebooks` anyway)

Reply:
Yeah but don't think it's problematic to do this in the forward pass, we'll have to shape & reshape anyways for generation, so IMO better if the forward pass can absorb the complexity. The forward pass of ConditionalGeneration is pretty much never used without generate anyways

Reply:
Okay sounds good - note that this reshaping should probably go in the forward call of `MusicgenForCausalLM` (the decoder only model), since we also want the generate method for this model to work standalone. E.g. in the example where you only want to fine-tune the decoder model, you also want to be able to evaluate the decoder only model by inputting token ids (prompts) and generating conditional audio samples, **without** having to construct the full `MusicgenForConditionalGeneration` model to assess performance

Reply:
Forward pass shapes resolved in https://github.com/huggingface/transformers/pull/24109/commits/a0cae3a55b38ab4b7036f6746f5fdef3477fafa2, docstrings in https://github.com/huggingface/transformers/pull/24109/commits/895a4594e0a80779d3b7cf632885343688e348d9